### PR TITLE
updated django backend and js ui with new purposes of visit.

### LINKED ISSUE #86: Request: Add More Purposes
Some users are visiting for different reasons we would like to track. Add these options:

- parts
- buy bike
- tour / visit

### PR DESCRIPTION
**Pull request by foltra**
Fixes #86 
Just a low hanging fruit PR, to get in the workflow.
Added a test for coverage.

### CODE REVIEW

**Comment by foltra** [comment]
Screen Shot 2019-06-12 at 4 50 06 PM

**Comment by drwlrsn** [comment]
Looks great. @FeXd I think these new options mean we really need to focus some effort on streamlining the _member_ creation flow.

File: bikeshop_project/assets/js/components/Purpose.jsx [hunk 1]
```diff
@@ -14,6 +14,9 @@ const Purpose = ({ initialValue, handleChange }) => (
     <MenuItem value={'WORKSHOP'} primaryText="Workshop" />
     <MenuItem value={'DONATE'} primaryText="Donate" />
     <MenuItem value={'STAFF'} primaryText="Staff" />
+    <MenuItem value={'PARTS'} primaryText="Parts" />
+    <MenuItem value={'BUY_BIKE'} primaryText="Buy Bike" />
+    <MenuItem value={'TOUR / VISIT'} primaryText="Tour / Visit" />
```
Comment (drwlrsn) [inline bikeshop_project/assets/js/components/Purpose.jsx]: Is this the right value? I kinda forget how this works... I have this feeling it's going to be `'TOUR'`

**Comment by foltra** [comment]
> I have just bit of a concern on the value we send to the API when we sign someone in for a tour. Could it be possible to write a test around this?

Yes, I was trying to make a test with a wrong value to fail but It didn't. Seems like `Visit.purpose` is not being validated on `visit_choices`, i.e. `clean()` is not being called?.

File: bikeshop_project/registration/views.py [hunk 2]
```diff
@@ -84,11 +85,14 @@ def dispatch(self, request, *args, **kwargs):
 
     def post(self, request):
         member = get_object_or_404(Member, id=request.POST.get('id'))
-        visit = signin_member(member, request.POST.get('purpose'))
         try:
+            visit = signin_member(member, request.POST.get('purpose'))
             membership = Membership.objects.select_related('payment').filter(member=member).last()
         except ObjectDoesNotExist:
             membership = None
+        except ValidationError:
```
Comment (foltra) [inline bikeshop_project/registration/views.py]: Decided to respond the request instead of just raising an exception

File: bikeshop_project/registration/forms.py [hunk 3]
```diff
@@ -56,3 +57,11 @@ def save(self, *args, **kwargs):
         if commit:
             instance.save()
         return instance
+
+class VisitForm(ModelForm):
```
Comment (foltra) [inline bikeshop_project/registration/forms.py]: Added this class here to use forms for validation

File: bikeshop_project/registration/utils.py [hunk 4]
```diff
@@ -21,7 +24,11 @@ def signin_member(member: Member, purpose: str) -> Visit:
     :raise: `AlreadySignedInError` or `ValidationError`
     """
     if not member_signed_in(member):
-        return Visit.objects.create(member=member, purpose=purpose)
+        instance = Visit.objects.create(member=member, purpose=purpose)
+        form = VisitForm(instance=instance, data={'member': member, 'purpose': purpose})
+        if form.is_valid():
+            return instance
+        raise ValidationError(instance)
```
Comment (drwlrsn) [inline bikeshop_project/registration/utils.py]: Did this not already raise a `ValidationError` or was the docblock lying?

Comment (foltra) [reply bikeshop_project/registration/utils.py]: Well, I wrote test for invalid purpose and It didn't fail as it should. I'm a bit rusty on django, but apparently `Visit.objects.create` doesn't validate fields or run `.clean()` .

**Comment by foltra** [comment]
image

**Comment by foltra** [comment]
Opened a new PR with base on branch  `heroku`  #96

File: bikeshop_project/registration/migrations/0006_auto_20190620_0334.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
```
Comment (drwlrsn) [inline bikeshop_project/registration/migrations/0006_auto_20190620_0334.py]: This is the same migration as https://github.com/BridgeCityBicycleCoop/workstand/pull/92/files/3dbe69723874e4778959545bf298b8c7eb2ea4de#diff-4cc176be3524646a52ccc48c1bd41820

I am going to merge the heroku branch first so we can do these subsequent PRs via a build pipeline with Heroku

Comment (foltra) [reply bikeshop_project/registration/migrations/0006_auto_20190620_0334.py]: removed a commit with the migration.

File: Pipfile [hunk 6]
```diff
@@ -52,7 +52,7 @@ requests = "*"
 coverage = "*"
 mock = "*"
 gunicorn = "==19.4.5"
-rollbar = "==0.13.11"
+rollbar = "==0.14.1"
```
Comment (drwlrsn) [inline Pipfile]: Let's do this in a separate commit.

File: Pipfile [hunk 7]
```diff
@@ -66,6 +66,7 @@ whitenoise = "*"
 algoliasearch = "==1.10.0"
 django-haystack-algolia = {editable = true,git = "https://github.com/drwlrsn/django-haystack-algolia.git",ref = "django-1.19"}
 #haystack-algolia = {editable = true,path = "./../django-haystack-algolia"} # Keep this around to update backend.
+django-waffle = "==0.16.0"
```
Comment (drwlrsn) [inline Pipfile]: Not sure what's going on, but this is already present in master.

**Comment by drwlrsn** [comment]
OH! This branch needs to be pointed at master.

**Comment by foltra** [comment]
Squashed into one commit and applyied to a fresh branch, best I could do to deal with the merge rebase spaghetti.
